[OSDOCS-10057]: Add TP docs about increasing DB size for etcd (#76718)
openshift/openshift-docs: modules/etcd-increase-db.adoc

// Module included in the following assemblies:
//
// * scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.adoc

:_mod-docs-content-type: PROCEDURE
[id="etcd-increase-db_{context}"]
= Increasing the database size for etcd

> You can set the disk quota in gigabytes for each etcd instance. If you set a disk quota for your etcd instance, the instance can increase up to the value that you specify.

Reviewer: ```suggestion
You can set the disk quota in gigabytes for each etcd instance. 
```

Reviewer: This sentence can be removed. The next sentences clarify that the value can be only increasing and set upto int 32.

You can specify integer values from 8 to 32. The default value is 8. You can specify only increasing values.

:FeatureName: Increasing the database size for etcd
include::snippets/technology-preview.adoc[]

[id="etcd-change-db-size_{context}"]
== Changing the etcd database size

> To change the database size for etcd, complete the following steps.

Reviewer: Is there a specific reason why users want to increase the size of the database, for example, to store the increasing workload/application data?

Up to you though!

Author: Good question!
@dusk125 What do you think of adding a line or two to explain why users might want to increase the size of their database? Or would that be obvious to most people who are reading these docs?

Reviewer 2: Yeah I think that's a good idea. Pretty much the only reasons would be
1. If they've experienced a ["Low Space" alert](https://github.com/openshift/cluster-etcd-operator/blob/master/manifests/0000_90_etcd-operator_03_prometheusrule.yaml#L90)

This alert is an indicator that they're cluster is getting too big to fit in etcd (assuming automatic-defrag is working properly) and they could benefit from this feature immediately; once etcd runs out of space, writes will start failing.

2. If they're experiencing ["Excessive Database Growth" alert (warning)](https://github.com/openshift/cluster-etcd-operator/blob/master/manifests/0000_90_etcd-operator_03_prometheusrule.yaml#L100)

This is an indicator that the database *may* get too big (but hasn't yet) and they may benefit from the feature and avoid the write-failing issue. This is a warning because it's just going off of the previous growth; if the growth stops, then they don't need this feature.. but if it does, following this warning could give them time to increase the size before running into write-failures.

Other than those reasons, I would suggest people not mess with it :D

.Procedure

> . Check to see what the current value is by entering the following command:

Author: Good suggestion!

+
[source,terminal]
----
$ oc describe etcd/cluster | grep "Backend Quota"
----
+
.Example output
[source,terminal]
----
Backend Quota Gi B: <value>
----

. Change the value by entering the following command:
+
[source,terminal]
----

> oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": <value>}}'

Reviewer: ```suggestion
$ oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": <value>}}'
```

----

. Review the output:
+
.Example output
[source,terminal]
----

> etcd.operator.openshift.io/cluster patched

Author: @dusk125 Just want to double check: Is this the correct output? I'd expect to see a value instead of a "cluster patched" message.

Reviewer 2: Will verify this today after my cluster starts up... but if memory serves, that should be correct.

Reviewer 2: Yeah there's no additional output other than the "patched" output.

ETCD_QUOTA_BACKEND_BYTES:                               8589934592
ETCD_QUOTA_BACKEND_BYTES:                               8589934592
ETCD_QUOTA_BACKEND_BYTES:                               8589934592
----

[id="etcd-change-db-size-troubleshooting_{context}"]
== Troubleshooting

If you encounter issues when you try to increase the database size for etcd, the following troubleshooting steps might help.

[id="etcd-troubleshoot-small-value_{context}"]
=== Value is too small

If the value that you specify is less than 8, you will see an error message. In the following example, the value of 5 is specified:

[source,terminal]
----
oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 5}}'
----

The output shows an error:

.Example output
[source,terminal]
----
The Etcd "cluster" is invalid: 
* spec.backendQuotaGiB: Invalid value: 5: spec.backendQuotaGiB in body should be greater than or equal to 8
* spec.backendQuotaGiB: Invalid value: "integer": etcd backendQuotaGiB may not be decreased
----

To resolve this issue, specify an integer between 8 and 32.

[id="etcd-troubleshoot-large-value_{context}"]
=== Value is too large

> If the value that you specify is greater than 32, you will see an error message. In the following example, the value of 64 is specified:

Reviewer: ```suggestion
If the value that you specify is greater than `32`, you see the following error message:
```

- Should 32 use back ticks because it is a valid value for `backendQuotaGiB`? Same question for all similar instances of values. 
- I think a present tense would suffice but up to you!
- "In the following example, the value of 64 is specified:" can be removed in favor of an example command

[source,terminal]
----
oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 64}}'
----

> 
> The output shows an error:
> 
> .Example output

Reviewer: ```suggestion
.Example error message
```

[source,terminal]
----
The Etcd "cluster" is invalid: spec.backendQuotaGiB: Invalid value: 64: spec.backendQuotaGiB in body should be less than or equal to 32
----

To resolve this issue, specify an integer between 8 and 32.

[id="etcd-troubleshoot-decreasing-value_{context}"]
=== Value is decreasing

> If the value is set to a valid value between 8 and 32 and you try to decrease the value, you will see an error message because the value can only increase. In the following example, the value is set to 10:

Reviewer: ```suggestion
If you set a valid value between 8 and 32, you cannot decrease the value. Otherwise, you see an error message.
```

[source,terminal]
----
oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 10}}'
----

The output shows that the change succeeded:

.Example output
[source,terminal]
----
etcd.operator.openshift.io/cluster patched
----

> After the rollout is completed, the value is decreased to 8:

Reviewer: Instead of setting a value as 10 and re-setting the value as 8, how about showing a current value as 10. And then setting a new value as 8.

Something like this:

--------------

1. Check to see what the current value is by entering the following command:

$ oc describe etcd/cluster | grep "Backend Quota"

Example output

Backend Quota Gi B: 10

2. Decrease the disk quota value by entering the following command:

oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 8}}'

.Example error message

The Etcd "cluster" is invalid: spec.backendQuotaGiB: Invalid value: "integer": etcd backendQuotaGiB may not be decreased

To resolve this issue, specify an integer greater than 10.

[source,terminal]
----

> oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 8}}'

Reviewer: ```suggestion
$ oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": 8}}'
```

a `$` sign is missing

----

> 
> The output shows an error:
> 
> .Example output

Reviewer: ```suggestion
.Example error message
```

:FeatureName: Increasing the database size for etcd
include::snippets/technology-preview.adoc[]

[id="etcd-change-db-size_{context}"]
== Changing the etcd database size

To change the database size for etcd, complete the following steps.

.Procedure

. Check the current value of the disk quota for each etcd instance by entering the following command:
+
[source,terminal]
----
$ oc describe etcd/cluster | grep "Backend Quota"
----
+
.Example output
[source,terminal]
----
Backend Quota Gi B: <value>
----

. Change the value of the disk quota by entering the following command:
+
[source,terminal]
----
$ oc patch etcd/cluster --type=merge -p '{"spec": {"backendQuotaGiB": <value>}}'
----
+
.Example output
[source,terminal]
----
etcd.operator.openshift.io/cluster patched
----

.Verification

. Verify that the new value for the disk quota is set by entering the following command:
+
[source,terminal]
----
$ oc describe etcd/cluster | grep "Backend Quota"
----
+
The etcd Operator automatically rolls out the etcd instances with the new values. 

> . Verify that the etcd pods are up and running by entering the following command:

Reviewer 2: `oc get pods -n openshift-etcd`
That'll show all of the etcd-related pods; specifically looking for the "4/4 Running" for each of the masters.

They could instead add a watch that would show the rollout happening instead of the instantaneous output of the above command.
`oc get pods -n openshift-etcd -w`

Whichever you think is better.

+
[source,terminal]
----

----
+
The following output shows the expected entries for master-0. 
+

> .Example output

Author: @dusk125 In the peer review, the reviewer suggested that we show the completed output instead of in progress output. Can you provide an example of what that might look like?

Reviewer 2: Will do this today, gotta startup a cluster to grab the output

Reviewer 2: @lahinson these are the entries for the full command

`oc get pods -n openshift-etcd`

```
NAME                                                   READY   STATUS      RESTARTS   AGE
etcd-ci-ln-b6kfsw2-72292-mzwbq-master-0                4/4     Running     0          39m
etcd-ci-ln-b6kfsw2-72292-mzwbq-master-1                4/4     Running     0          37m
etcd-ci-ln-b6kfsw2-72292-mzwbq-master-2                4/4     Running     0          41m
etcd-guard-ci-ln-b6kfsw2-72292-mzwbq-master-0          1/1     Running     0          51m
etcd-guard-ci-ln-b6kfsw2-72292-mzwbq-master-1          1/1     Running     0          49m
etcd-guard-ci-ln-b6kfsw2-72292-mzwbq-master-2          1/1     Running     0          54m
installer-5-ci-ln-b6kfsw2-72292-mzwbq-master-1         0/1     Completed   0          51m
installer-7-ci-ln-b6kfsw2-72292-mzwbq-master-0         0/1     Completed   0          46m
installer-7-ci-ln-b6kfsw2-72292-mzwbq-master-1         0/1     Completed   0          44m
installer-7-ci-ln-b6kfsw2-72292-mzwbq-master-2         0/1     Completed   0          49m
installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-0         0/1     Completed   0          40m
installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-1         0/1     Completed   0          38m
installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-2         0/1     Completed   0          42m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-0   0/1     Completed   0          43m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-1   0/1     Completed   0          43m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-2   0/1     Completed   0          43m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-0   0/1     Completed   0          42m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-1   0/1     Completed   0          42m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-2   0/1     Completed   0          42m
```

// Module included in the following assemblies:
//
// * scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.adoc

:_mod-docs-content-type: PROCEDURE
[id="etcd-increase-db_{context}"]
= Increasing the database size for etcd

You can set the disk quota in gigabytes for each etcd instance. If you set a disk quota for your etcd instance, you can specify integer values from 8 to 32. The default value is 8. You can specify only increasing values.

> 

Reviewer 2: Could we add something along the lines of:
Increasing this value does not immediately reserve this amount of disk space, but instead allows etcd to grow up to that value if it needs to. Please ensure that etcd is running on a dedicated disk larger than the value you wish to set.

(Not sure if we need to be more specific about how much larger the disk needs to be)

Author: Done. See line 15.

> You might want to increase the disk quota if you encounter a `low space` alert. This alert indicates that the cluster is too large to fit in etcd, assuming that automatic defragmentation is working properly. If you see this alert, you need to increase the disk quota immediately because after etcd runs out of space, writes fail.

Author: Revised as you suggested

installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-0         0/1     Completed   0          40m
installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-1         0/1     Completed   0          38m
installer-8-ci-ln-b6kfsw2-72292-mzwbq-master-2         0/1     Completed   0          42m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-0   0/1     Completed   0          43m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-1   0/1     Completed   0          43m
revision-pruner-7-ci-ln-b6kfsw2-72292-mzwbq-master-2   0/1     Completed   0          43m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-0   0/1     Completed   0          42m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-1   0/1     Completed   0          42m
revision-pruner-8-ci-ln-b6kfsw2-72292-mzwbq-master-2   0/1     Completed   0          42m
----

. Verify that the disk quota value is updated for the etcd pod by entering the following command:
+
[source,terminal]
----
$ oc describe -n openshift-etcd pod/<etcd_podname> | grep "ETCD_QUOTA_BACKEND_BYTES"
----
+
The value might not have changed from the default value of `8`.
+
.Example output
[source,terminal]
----

> ETCD_QUOTA_BACKEND_BYTES:                               8589934592

Author: Done. I added a note after the output example.
